yt_dlp/extractor/common.py

```diff
@@ -1854,12 +1854,24 @@ def _check_formats(self, formats, video_id):
 
     @staticmethod
     def _remove_duplicate_formats(formats):
-        format_urls = set()
+        seen_urls = set()
+        seen_fragments = set()
         unique_formats = []
         for f in formats:
-            if f['url'] not in format_urls:
-                format_urls.add(f['url'])
+            fragments = f.get('fragments')
+            if fragments and not callable(fragments):
+                if base_url := f.get('fragment_base_url'):
+                    fragments = map(urljoin(base_url), fragments)
+
+                fragments = frozenset(fragments)
+                if fragments not in seen_fragments:
+                    seen_fragments.add(fragments)
+                    unique_formats.append(f)
+
+            elif f['url'] not in seen_urls:
+                seen_urls.add(f['url'])
                 unique_formats.append(f)
```

Comment:
So if a format has `fragments` but it is a generator/lazy, then we evaluate uniqueness by `url` instead?

Reply:
would it be better to just always add the format if `fragments` is callable?
```suggestion
            if fragments:
                if callable(fragments):
                    unique_formats.append(f)
                    continue
                if base_url := f.get('fragment_base_url'):
                    fragments = map(urljoin(base_url), fragments)

                fragments = frozenset(fragments)
                if fragments not in seen_fragments:
                    seen_fragments.add(fragments)
                    unique_formats.append(f)

            elif f['url'] not in seen_urls:
                seen_urls.add(f['url'])
                unique_formats.append(f)
```